### PR TITLE
Input rework

### PR DESCRIPTION
**Pull request by artumino**
I'll open this one to discuss possible changes to inputs, since it's a delicate topic It'll probably stay as draft for a long time.

**Suggestions that were tossed around**

- [x] Allow tool controls to be bound to different buttons for controllers with buttons to spare (index controllers and original wmr controllers for example) **[READY]**
- [ ] Make additional cockpit buttons to operate Auto-Nav and Tools while in the ship **[FuturePR]**
   - [ ] Laser-pointer interactions
   - [ ] Physical interaction? (why not, if we place them near the arm rests they could be in reach)
   - [ ] Visual Feedback (light them up appropriatly)
- [ ] Enable touch controls for hand-held tools to allow freedom of movement for controllers with less buttons **[FuturePR]**
- [ ] Select items in the shiplog by laser-pointer  **[FuturePR]**
- [x] Hand independent interactions **[READY]**
   - [x] Toggle light with both hands
   - [x] Pickup tools and operate them with either hand

**Things that might change with the new system**

- I have updated the SteamVR plugin version in this branch to the latest one, if issues arise it should be pretty straight forward to revert the change.
- ProxymityDetectors right now are always updating, would using the proper Unity collision system with triggers and kinematik rigidbodies on the hands break everything? The number of detectors might increase a bit if we want to have physical interactions with the tools and cockpit buttons.
- #390 Might need to be merged before this one, a lot of the interactions proposed would need to use custom poses

### CODE REVIEW

**Comment by Raicuparta** [comment]
For the cockpit controls I'd prefer to use laser controls with buttons in the screens we already use for those. Would mean an easier change, and less trouble for people playing in restricted areas.

**Comment by Raicuparta** [comment]
For the handheld tools I don't mind actual direct touch interaction since the player will be in full control of the touchable buttons (primary hand) while touching them with their off-hand.

**Comment by Raicuparta** [comment]
And most of this can be split in smaller PRs, but you probably already intend to do that.

**Comment by Raicuparta** [comment]
I did have to make some changes to the SteamVR code at some point. I think I left some comments indicating where I made changes. Did you notice that? Might be just one place, don't remember if I reverted the others.

**Comment by Raicuparta** [comment]
The SteamVR update by itself should be an individual PR. Otherwise all these changes are gonna be really hard to review.

**Comment by artumino** [comment]
As of d7316c7 hand independent interactions and different bindings while holding tools should be functioning as expected.
(I still need to create default bindings for oculus, wmr and vive users)

There are a few issues to address (some of them will be addressed with the other proposed changes):
- The translator page lights (the arrows near the page numbers) are mirrored on the left hand
- Holding items in the left hand disables movement on controllers that have not enough buttons (will be mitigated once we provide a mirrored set of controls for left hand users, but can still be annoying for right handed users)
- The button prompts are still tied to the main hand (the one holding the laser pointer)

**Comment by artumino** [comment]
> 
> 
> The SteamVR update by itself should be an individual PR. Otherwise all these changes are gonna be really hard to review.

As for this, I've moved it to a separate branch and will open a PR, I don't think it's necessary to update actually even though there's a small change that might save up some allocations on the pose update loop...

**Comment by artumino** [comment]
Ok, I think testing for the current open changes can start. 

The current open issues that I'm aware of are:
- When picking up a tool on controllers that override the move/rotate axes those axes are not reset to 0 so the player keeps moving/rotating until the tool is unequipped (this might also be already present in the current release)
- I'm not quite satisfied with how the concept of hand dominance is spread between multiple components instead of being in a signle one, I'll have to refactor things around

Things that I also wanted to investigate:
- In the current release sometimes the hold to repair button isn't working, this might be a good time to fix it
- I wanted to compare performance with previous releases to be sure that rebuilding prompts doesn't cause any issues

**Comment by artumino** [comment]
@Raicuparta did you manage to get the VR setup going again?

**Comment by Raicuparta** [comment]
> @Raicuparta did you manage to get the VR setup going again?

Yes I did. Lemme know when you want me to test stuff.

**Comment by artumino** [comment]
Ok, If you want to give it a go try it 👍

The only thing that is left is to flip pickup items when holding them with left as dominant hand.

File: NomaiVR/Input/ControllerInput.cs [hunk 1]
```diff
@@ -249,6 +249,14 @@ private void RegisterToActiveBindingChanged(ISteamVR_Action_In actionIn, Action<
             private void OnBaseBindingChanged(ISteamVR_Action fromAction, SteamVR_Input_Sources fromSource, bool active)
             {
                 if(active && fromAction != null && fromAction.active) _baseBindingsChanged = true;
+
```
Comment (artumino) [inline NomaiVR/Input/ControllerInput.cs:252]: Just realized this should be done only once, leaving this for a future change

**Comment by artumino** [comment]
Ok, the only thing missing for merging this one is the WMR bindings.
I need to find someone with a G2 and the original controllers and redo the bindings...

**Comment by Raicuparta** [comment]
Tested, seems perfect to me. I can't look too deeply into the code but I trust that it's fine :P Do I understand that you made some changes to how the prompts are generated? To make them re-generate when needed?

There's one big problem with the NomaiVR prompts. You may have had the game crash on you before if you didn't have the controllers connected when starting the game. This is because I purposely force crash the game if one or both controllers aren't detected. Because if controllers aren't there, I can't generate the prompt messages correctly, and at the time I couldn't figure out a good way to delay the prompt text generation until later (it seemed like I could only do it when the prompt is created). Maybe some adjustments to your implementation could help with this? These forced crashes are problematic for a lot of people first trying the mod. For another PR of course. We can discuss it if you're interested.

**Review by Raicuparta — APPROVED** [review]

**Comment by artumino** [comment]
Yes, I hooked up the prompt rebuilding to active bindings changes. So whenever an action is enabled on a new controller it rebuilds all the prompts that are dependent on that action if they changed. This should fix the issue of prompts without any button in them from happening, it also allows the prompts to rebuild if a button gets remaped without restarting the game.

**Comment by Raicuparta** [comment]
I noticed that it worked better yeah. But I have this thing that generates the text and adds "left hand / right hand" to the prompt, but only if there is a button with the same name in the opposite hand. So if you start the game with only one controller active, this algorithm fails to detect that. So when you refresh the prompts, it needs to refresh all of them to make sure the text is correct. Not sure if that is clear.

**Comment by artumino** [comment]
I wrote that incorrectly, If it is a base binding it rebuilds all of them again, so the ones that have the Left/Right hand logic are rebuilt for both controllers. I selectively rebuild prompts only for bindings that are always tied to one hand (so for example the tool bindings).

**Comment by Raicuparta** [comment]
Not sure if I understand. But I tested in your branch by starting the game with only one controller, and the binding text was incorrect, it didn't add the "left hand" text. I think you need always rebuild all prompts no matter what, because it needs to go through all of them to check if the names need to be disambiguated. For instance, the joysticks are just called "stick", so they need the hand text to be added. So if the left controller is connected later, it should redo the prompt for the right controller too.

**Comment by artumino** [comment]
@Raicuparta give it a try now, the prompts were rebuilt but there was a bug elsewhere that prevented the hand and click part to update properly

**Pull request merged**